src/ocean/meta/values/VisitValue.d

```diff
@@ -0,0 +1,155 @@
+        Alternatively, this file may be distributed under the terms of the Tango
+        3-Clause BSD License (see LICENSE_BSD.txt for details).
+
+*******************************************************************************/
+
+module ocean.meta.values.VisitValue;
+
+import ocean.meta.traits.Basic;
+import ocean.meta.types.Arrays;
+import ocean.meta.types.Typedef;
+
+/*******************************************************************************
+
+    Recursively iterates over `value` fields and calls `visitor.visit` for each
+    one of those. `visitor.visit` is expected to be a template with the
+    following signature:
+
+    `bool visit ( T ) ( T* value )`
+
+    The return value indicates whether the value should be recursed into or
+    not.
+
+    Example:
+
+    ---
+    struct SumIntegersVisitor
+    {
+        /// Visitor can have any internal state it may need
+        long sum;
+
+        /// Returns: 'true' if recursion shall continue for this type
+        bool visit ( T ) ( T* value )
+        {
+            static if (isIntegerType!(T))
+                sum += *value;
+
+            return true;
+        }
+    }
+
+    SumIntegersVisitor visitor;
+    visitValue(some_value, visitor);
+    ---
+
+    Params:
+        value = root value to visit
+        visitor = struct providing templated `visit` method accepting single
+            argument which will be a pointer to visited element
+
+*******************************************************************************/
+
+public void visitValue ( T, Visitor ) ( ref T value, ref Visitor visitor )
+{
+    VisitImpl!(Visitor) impl;
+    impl.visitor = &visitor;
+    impl.visitAll(&value);
+}
+
+/*******************************************************************************
+
+    Implementation of visiting logic. Some types are currently not supported
+    because it is not clear what would recursing mean for them.
+
+*******************************************************************************/
+
+private struct VisitImpl ( Visitor )
+{
+    // Refer to the persistent instance of visitor struct in case implementation
+    // may need internal state
+    Visitor* visitor;
+
+    // Main recursive visiting implementation
+    private void visitAll ( T ) ( T* value )
+    {
+        auto deeper = this.visitor.visit(value);
+
+        if (!deeper)
+            return;
+
+        static if (isPrimitiveType!(T))
+        {
+            // do nothing, already processed
+        }
+        else static if (isTypedef!(T))
+        {
+            auto reinterp = cast(TypedefBaseType!(T)*) value;
+            this.visitAll(reinterp);
+        }
+        else static if (isAggregateType!(T))
+        {
+            static if (is(T == class))
+                bool ignore = *value is null;
+            else
+                bool ignore = false;
+
+            if (!ignore)
+            {
+                foreach (ref field; (*value).tupleof)
```

Comment:
What about super class fields? e.g. https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/util/DeepReset.d#L271-L281

Reply:
Good catch! I should add tests for that too